src/index.js

```diff
@@ -1,118 +1,157 @@
 
     return next => action => {
-      let promise = null;
-      let data = null;
 
+      /**
+       * Instantiate variables to hold:
+       * (1) the promise
+       * (2) the data for optimistic updates
+       */
+      let promise;
+      let data;
+
+      /**
+       * There are multiple ways to dispatch a promise. The first step is to
+       * determine if the promise is defined:
+       * (a) explicitly (action.payload.promise is the promise)
+       * (b) implicitly (action.payload is the promise)
+       * (c) as an async function (returns a promise when called)
+       *
+       * If the promise is not defined in one of these three ways, we don't do
+       * anything and move on to the next middleware in the middleware chain.
+       */
+
+      // Step 1a: Is there a payload?
       if (action.payload) {
-        data = action.payload.data;
+        const PAYLOAD = action.payload;
 
-        // If there is a payload and it has a promise attribute, we'll try
-        // that
-        if (action.payload.promise) {
-          promise = action.payload.promise;
+        // Step 1.1: Is the promise implicitly defined?
+        if (isPromise(PAYLOAD)) {
+          promise = PAYLOAD;
+        }
 
-        // Otherwise we'll try the payload
-        } else {
-          promise = action.payload;
+        // Step 1.2: Is the promise explicitly deifned?
+        else if (isPromise(PAYLOAD.promise)) {
+          promise = PAYLOAD.promise;
+          data = PAYLOAD.data;
         }
-      // If there's no payload just return early.
-      } else {
-        return next(action);
-      }
 
-      // If the promise we're tracking is a regular function, call it and see if
-      // it returns a promise.
-      if (typeof promise === 'function') {
-        const functionResult = promise();
-
-        // If it is a promise, awesome, let's use it.
-        if (isPromise(functionResult)) {
-          promise = functionResult;
-
-        // Return early if otherwise. We might be messing with some other redux
-        // middleware at this point.
-        // TODO Should we always pass `action` or `functionResult` to `next`? Or
-        // keep what's here.
-        } else {
-          return next({
-            type: action.type,
-            payload: functionResult,
-            ...(action.meta !== undefined ? action.meta : {})
-          });
+        // Step 1.3: Is the promise returned by an async function?
+        else if (
+          typeof PAYLOAD === 'function' ||
+          typeof PAYLOAD.promise === 'function'
+        ) {
+          promise = PAYLOAD.promise ? PAYLOAD.promise() : PAYLOAD();
+          data = PAYLOAD.promise ? PAYLOAD.data : undefined;
+
+          // Step 1.3.1: Is the return of action.payload a promise?
+          if (!isPromise(promise)) {
+
+            // If not, move on to the next middleware.
+            return next({
+              ...action,
+              payload: promise
+            });
```

Comment:
Is `action.meta` still correctly passed here?

Reply:
Yes, and I added a test for it: https://github.com/pburtchaell/redux-promise-middleware/pull/184/commits/7ceb35e86d2da36c71f7360b171c2e046d3ed81d